### PR TITLE
Wrote flutter_todos tutorial (for v8)

### PR DESCRIPTION
**Pull request by chonghorizons**
Wrote the flutter_todos tutorial. First draft. It is complete except that I didn't generate the video and there are some comment fields to remove.
Completes Issue #2859

NOTE: This merge is into the `felangel:docs/flutter-todos-v8.0.0` branch. Not a merge to the master branch. (Did I do this right?)

## Status

Ready for review of text. Then, two final steps.

## Breaking Changes

NO

## Description

Tasks
[x] Removed old MD code snippets
[x] Made MD files for every dart file
[x] Inserted them into the tutorial docs, ensuring complete file coverage
[x] Made widget tree diagrams
[x] Added an Architecture section
[x] Consolidated many bits of commentary into an "event flow" section.
[ ] first review by reviewers
[ ] incorporate feedback
[ ] go through all the ZTODO instances and address those. Remove any commenting tokens I added.
[ ] make the gif at the beginning.
[ ] merge and pull request to main
[ ] celebrate

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [ ] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [x] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

**Review by chonghorizons — COMMENTED** [review]
This commit c53c4de has the most files. It's a bulk delete of the old snippets and an add of the new snippets.

File: docs/_snippets/flutter_todos_tutorial/lib/app.dart.md [hunk 1]
```diff
@@ -0,0 +1,38 @@
+```dart
+import 'package:flutter/material.dart';
```
Comment (felangel) [inline docs/_snippets/flutter_todos_tutorial/lib/app.dart.md:2]: Where possible, I would highly recommend/prefer to avoid having snippets which are identical to the actual implementation. In those cases, we can just reference the actual files in GitHub to avoid duplication and to ensure the snippets are always in sync and up to date. Let me know if that makes sense or if you have any questions, thanks!

File: docs/fluttertodostutorial.md [hunk 2]
```diff
@@ -1,483 +1,1488 @@
+
```
Comment (felangel) [inline docs/fluttertodostutorial.md:1]:
```suggestion
```
nit but can we remove the extra newline?

File: docs/fluttertodostutorial.md [hunk 3]
```diff
@@ -1,483 +1,1488 @@
...
+> ZZTODO: Update gif with the movie in the issue tracker.
 
 ## Key Topics
 
 - Observe state changes with [BlocObserver](/coreconcepts?id=blocobserver).
 - [BlocProvider](/flutterbloccoreconcepts?id=blocprovider), Flutter widget which provides a bloc to its children.
 - [BlocBuilder](/flutterbloccoreconcepts?id=blocbuilder), Flutter widget that handles building the widget in response to new states.
-- Using Bloc instead of Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)
+- [BlocListener](/flutterbloccoreconcepts?id=bloclistener), a Flutter widget which invokes the listener code in response to state changes in the bloc.
+- Using 3 Blocs and  1 Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)
```
Comment (felangel) [inline docs/fluttertodostutorial.md:17]:
```suggestion
- Using Blocs and Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)
```

I feel the amount isn't that important and it could get outdated pretty quickly. wdyt?

File: docs/fluttertodostutorial.md [hunk 4]
```diff
@@ -1,483 +1,1488 @@
...
+- Uses a [Repository and API layer](architecture/repository) separately for the data.
+- [RespositoryProvider](/flutterbloccoreconcepts?id=respositoryprovider), a flutter widget to provide a repository to its children.
 - Prevent unnecessary rebuilds with [Equatable](/faqs?id=when-to-use-equatable).
-- [MultiBlocProvider](/flutterbloccoreconcepts?id=multiblocprovider), a Flutter widget that merges multiple BlocProvider widgets into one.
+- [MultiBlocListener](/flutterbloccoreconcepts?id=multibloclistener), a Flutter widget that merges multiple BlocListener widgets into one.
+
```
Comment (felangel) [inline docs/fluttertodostutorial.md:22]:
```suggestion
```

File: docs/fluttertodostutorial.md [hunk 5]
```diff
@@ -1,483 +1,1488 @@
...
 
 ## Setup
 
 We'll start off by creating a brand new Flutter project
 
-[script](_snippets/flutter_todos_tutorial/flutter_create.sh.md ':include')
+```shell
+flutter create flutter_login
```
Comment (felangel) [inline docs/fluttertodostutorial.md:29]: Can we revert this change? The previous script should still be correct and it should be `flutter create flutter_todos` not `flutter_login`.

File: docs/fluttertodostutorial.md [hunk 6]
```diff
@@ -1,483 +1,1488 @@
...
-- [MultiBlocProvider](/flutterbloccoreconcepts?id=multiblocprovider), a Flutter widget that merges multiple BlocProvider widgets into one.
+- [MultiBlocListener](/flutterbloccoreconcepts?id=multibloclistener), a Flutter widget that merges multiple BlocListener widgets into one.
+
 
 ## Setup
 
 We'll start off by creating a brand new Flutter project
 
-[script](_snippets/flutter_todos_tutorial/flutter_create.sh.md ':include')
+```shell
```
Comment (felangel) [inline docs/fluttertodostutorial.md:28]:
```suggestion
```sh
```

minor but for consistency can we just `sh`?

File: docs/fluttertodostutorial.md [hunk 7]
```diff
@@ -1,483 +1,1488 @@
...
 [pubspec.yaml](_snippets/flutter_todos_tutorial/pubspec.yaml.md ':include')
 
 and then install all of the dependencies
 
-[script](_snippets/flutter_todos_tutorial/flutter_packages_get.sh.md ':include')
+```shell
+flutter pacakages get
+```
+
+[comment]: <> (> ZTODO: Is below still true? I think so, but I don't know the history of this.)
```
Comment (felangel) [inline docs/fluttertodostutorial.md:42]: No this shouldn't be the case anymore with the rewritten version.

File: docs/fluttertodostutorial.md [hunk 8]
```diff
@@ -1,483 +1,1488 @@
...
+```
 
 We can then replace the contents of `pubspec.yaml` with
 
 [pubspec.yaml](_snippets/flutter_todos_tutorial/pubspec.yaml.md ':include')
 
 and then install all of the dependencies
 
-[script](_snippets/flutter_todos_tutorial/flutter_packages_get.sh.md ':include')
+```shell
```
Comment (felangel) [inline docs/fluttertodostutorial.md:38]: Can we revert this change as well?

File: docs/_snippets/flutter_todos_tutorial/lib/app.dart.md (hunk 1 above)
Comment (chonghorizons) [reply docs/_snippets/flutter_todos_tutorial/lib/app.dart.md:2]: 100% Agree, with comment.

- Linking to the tutorial on `master` syncs to master, but the tutorial text might reflect an older version.
- Making a snippet makes sure the website has a current snapshot at the time of writing. I've seen this happen for meteor and nextJS docs after major refactors.
  -  **What do you think about linking to a specific commit?** (then, on any tutorial update, one can find/replace to point to the new commit URL. So one would be deliberate in updating the version of linked code files.)
- I was following the use of snippers of what I saw in other tutorials. I suggested an edit to `Contributing` in the wiki. See #3056. I'm not wed to the language.

Recommend not letting this be blocking. This can be a later re-factor. It's probably about 1 hour of work.

File: docs/fluttertodostutorial.md (hunk 3 above)
Comment (chonghorizons) [reply docs/fluttertodostutorial.md:17]: I could go either way, but I like keeping the numbers. Otherwise, the stuff at the top is easy to skip over.

File: docs/fluttertodostutorial.md (hunk 4 above)
Comment (chonghorizons) [reply docs/fluttertodostutorial.md:22]: Good catch.

**Comment by chonghorizons** [comment]
Keeping track of a list of final polish before merge to master:

- convert "\`\`\`shell" to "\`\`\`sh"

File: docs/fluttertodostutorial.md (hunk 8 above)
Comment (chonghorizons) [reply docs/fluttertodostutorial.md:38]: Yes. This might be faster. I recommend making a list of final polish/styleguide steps here:

https://github.com/felangel/bloc/pull/3047#issuecomment-989334851

If you focus on formatting you'll never get through the tutorial.

It's faster to just write a list of style-guide items to implement (like a human-implemented css file) than it is to make the individual changes.

Take a look via docsify. There are notes to ignore in the raw markdown.

File: docs/_snippets/flutter_todos_tutorial/lib/app.dart.md (hunk 1 above)
Comment (marcossevilla) [reply docs/_snippets/flutter_todos_tutorial/lib/app.dart.md:2]: I think we can do a merge of the tutorial’s code branch and then reference the master branch with the new files, so @chonghorizons doesn’t need to change the files location when we merge this PR, in case there are a lot of entire files being used as snippets, let me know what you both think 👍

Comment (marcossevilla) [reply docs/_snippets/flutter_todos_tutorial/lib/app.dart.md:2]: otherwise we can simplify the snippets with the chunk of code we’re referencing instead of pasting the whole file

File: docs/fluttertodostutorial.md [hunk 9]
```diff
@@ -1,483 +1,1488 @@
...
 
 ## Key Topics
 
 - Observe state changes with [BlocObserver](/coreconcepts?id=blocobserver).
 - [BlocProvider](/flutterbloccoreconcepts?id=blocprovider), Flutter widget which provides a bloc to its children.
 - [BlocBuilder](/flutterbloccoreconcepts?id=blocbuilder), Flutter widget that handles building the widget in response to new states.
-- Using Bloc instead of Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)
+- [BlocListener](/flutterbloccoreconcepts?id=bloclistener), a Flutter widget which invokes the listener code in response to state changes in the bloc.
+- Using 3 Blocs and  1 Cubit. [What's the difference?](/coreconcepts?id=cubit-vs-bloc)
+- Uses a [Repository and API layer](architecture/repository) separately for the data.
```
Comment (marcossevilla) [inline docs/fluttertodostutorial.md:18]:
```suggestion
- Uses a [Repository and API layer](architecture/repository) separately for the domain and data logic, respectively.
```

File: docs/fluttertodostutorial.md [hunk 10]
```diff
@@ -1,483 +1,1488 @@
...
+- [RespositoryProvider](/flutterbloccoreconcepts?id=respositoryprovider), a flutter widget to provide a repository to its children.
```
Comment (marcossevilla) [inline docs/fluttertodostutorial.md:19]:
```suggestion
- [RepositoryProvider](/flutterbloccoreconcepts?id=respositoryprovider), a Flutter widget which provides a repository to its children.
```

File: docs/fluttertodostutorial.md [hunk 11]
```diff
@@ -1,483 +1,1488 @@
...
 - Prevent unnecessary rebuilds with [Equatable](/faqs?id=when-to-use-equatable).
-- [MultiBlocProvider](/flutterbloccoreconcepts?id=multiblocprovider), a Flutter widget that merges multiple BlocProvider widgets into one.
+- [MultiBlocListener](/flutterbloccoreconcepts?id=multibloclistener), a Flutter widget that merges multiple BlocListener widgets into one.
```
Comment (marcossevilla) [inline docs/fluttertodostutorial.md:21]:
```suggestion
- [MultiBlocListener](/flutterbloccoreconcepts?id=multibloclistener), a Flutter widget that sets up multiple BlocListener widgets that allow us to react to different states.
```

Comment (marcossevilla) [reply docs/fluttertodostutorial.md:21]: Not sure about this one, it can be reformed better

File: docs/fluttertodostutorial.md (hunk 5 above)
Comment (marcossevilla) [reply docs/fluttertodostutorial.md:29]: I think we should mention that the project was built using `very_good_cli` to redirect them to that package documentation in case they have questions about specific features that the CLI sets for the project, wdyt?

File: docs/fluttertodostutorial.md [hunk 12]
```diff
@@ -1,483 +1,1488 @@
...
+[comment]: <> (> ZTODO: Is below still true? I think so, but I don't know the history of this.)
+[comment]: <> (> **Note:** We're overriding some dependencies because we're going to be reusing them from [Brian Egan's Flutter Architecture Samples]&#40;https://github.com/brianegan/flutter_architecture_samples&#41;.)
+
+## Directory Structure
+
+Because this is a long tutorial, presenting the directory structure is helpful.
+
+> Alternatively, explore the files on github or clone them locally and open in your IDE. The full source for this example can be found [here](https://github.com/felangel/Bloc/tree/master/examples/flutter_todos).
+
+We will be using 3 blocs, 1 cubit, and 3 packages.
```
Comment (marcossevilla) [inline docs/fluttertodostutorial.md:51]: I think we should explain this in sections, @chonghorizons I can share with you the outline I had for this tutorial so we can just explain the components future-proofing the quantities we’re using right now in case we make the tutorial more complex, we just need to add a new section and not changing these numbers, wdyt?

**Review by marcossevilla — COMMENTED** [review]
I have some initial comments and overall tutorial looks good but I think we can restructure the content. By that, I don’t mean the content is wrong, but to reorder it and setting an outline for it to be easier to read and “modularizing” it for future edits we may do if we want to extend the tutorial, looking forward to read your thoughts 👍

File: docs/fluttertodostutorial.md (hunk 11 above)
Comment (chonghorizons) [reply docs/fluttertodostutorial.md:21]: Let's change to:
... sets up ...... at once. OR
... sets up ...... at the same point in the widget tree.

The "allows us to react to different states" isn't unique to MultiBlocListener. It's just a convenience /shortcut.

BTW: can you @marcossevilla add me as a collaborator on this issue or ask felix to do so? Otherwise, I can't suggest the change directly, I think, in github. 

Add suggestion to batch is greyed out.

File: docs/fluttertodostutorial.md (hunk 12 above)
Comment (chonghorizons) [reply docs/fluttertodostutorial.md:51]: That would be okay, but I think the numbers are helpful for newcomers reading the docs. It's also pointing out to look at the directory structure to see this.

It could be more clear by breaking it out as a

> TIP: Look at the directory structure. By using a good naming convention and directory structure, we can easily see: We will be using 3 blocs, 1 cubit, and 3 packages.

Sorry, I can't edit directly since I'm not a collaborator.

File: docs/fluttertodostutorial.md (hunk 11 above)
Comment (felangel) [reply docs/fluttertodostutorial.md:21]: @chonghorizons you should be able to edit because it's your branch.

**Comment by felangel** [comment]
@chonghorizons FYI I'm planning to go through this in the next few days and will apply the feedback and merge it into #2725  by this weekend 👍

**Comment by chonghorizons** [comment]
@felangel cool.  Make any changes you want without needing my approval. Including big cuts. I trust your judgment and am not possessive.

Hope it helps get this great tutorial out!

**Comment by chonghorizons** [comment]
> I have some initial comments and overall tutorial looks good but I think we can restructure the content. By that, I don’t mean the content is wrong, but to reorder it and setting an outline for it to be easier to read and “modularizing” it for future edits we may do if we want to extend the tutorial, looking forward to read your thoughts 👍

Recommend we make a separate doc issue of 
- making a style guide
- modularizing

Important, but not sure it should block the initial update.

File: docs/fluttertodostutorial.md (hunk 11 above)
Comment (chonghorizons) [reply docs/fluttertodostutorial.md:21]: @felangel Unfortunately, "Add suggestion to branch" is still greyed out:
image

I can edit the forked branch, but I think the issue is that I don't have write permission on this branch.

**Comment by marcossevilla** [comment]
> Recommend we make a separate doc issue of
> 
> * making a style guide
> * modularizing
> 
> Important, but not sure it should block the initial update.

yeah totally agree, I'm going to work on restructuring your latest changes this weekend so I'll probably be uploading a PR for your branch soon if that's ok

**Comment by chonghorizons** [comment]
> yeah totally agree, I'm going to work on restructuring your latest changes this weekend so I'll probably be uploading a PR for your branch soon if that's ok

@marcossevilla , Go for it!

**Comment by felangel** [comment]
@marcossevilla are you working on this? I'd like to try to get this merged this weekend so if you're busy no worries, I can try to take it over. Thanks! 🙏

**Comment by marcossevilla** [comment]
> @marcossevilla are you working on this? I'd like to try to get this merged this weekend so if you're busy no worries, I can try to take it over. Thanks! 🙏

yup, probably leaving the PR up today 👍

**Comment by felangel** [comment]
> > @marcossevilla are you working on this? I'd like to try to get this merged this weekend so if you're busy no worries, I can try to take it over. Thanks! 🙏
> 
> yup, probably leaving the PR up today 👍

Awesome let me know if you need any help, thanks! 💙

**Comment by marcossevilla** [comment]
@felangel @chonghorizons PR is up! You can check it [here](https://github.com/chonghorizons/bloc/pull/1). It shouldn't be considered as approval for this PR, rather as a restructure of the content itself, so we can re-add feedback with a more comprehensible and extensible write-down. 👍

**Comment by felangel** [comment]
Going to merge this into the todos v8 branch and will go through and make any final adjustments and try to publish within the next day or two. Thanks so much @chonghorizons and @marcossevilla! 💙 🙏

**Comment by chonghorizons** [comment]
@felangel That's fast! Cool.

There is one section that @marcossevilla removed/heavily modified that you won't see if you look at what was merged. That was a new **Architecture** section that presented the widget tree early on. 

I'll put another pull request in so you can see it and then decide whether to include or remove.

The discord and issues/GitHub get lots of architecture questions.

Including means that people can be pointed to "look at the architecture discussion section of flutter todos" to see how it's done when they ask those questions.

Not including means there isn't any explicit discussion of architecture in any tutorial. (There is the architecture discussion in the package section of the docs, but this isn't discussed *as applied* in any tutorial, afaik)

**Comment by marcossevilla** [comment]
@chonghorizons I removed it because IMO we should have all the architecture questions addressed on [its own section](https://bloclibrary.dev/#/architecture). That way we can specify architectural patterns from every example and link to those there.

If we set that explanation in one example, people will have to check out that example to get the explanation instead of checking the dedicated section to architecture on the docs, which is confusing IMO.

**Comment by felangel** [comment]
I agree with @marcossevilla that ideally the architecture section of the docs should cover core architectural concepts and the rationale. I’m working on revamping the architecture docs and will have an update shortly.

**Comment by chonghorizons** [comment]
@marcossevilla , @felangel 

Sounds better & good to me to have a common architecture section that is general and that each tutorial can link to.

Do we want individual tutorials to have an architectural section? 
> If not, we perpetuate the (mistaken?) idea that bloc is ~100% state management. This example that @jeroen-meijer wrote demonstrates that bloc can be very important for architecture. Architecture is maybe 25-50% in this example.

The tutorial section might be something like:
---
### Architecture
* See https://bloclibrary.dev/#/architecture for a review of architectural considerations. **BOILERPLATE**
* In this tutorial, these specific architectural decisions were made: **PROJECT SPECIFIC**
  * xxx
  * yyy
  * zzz
  * One Repo, 3 Blocs, One Bloc per feature/view, BlocProviders as low as possible.

#### Architecture, deeper discussion
In this case, 50% of using `flutter_bloc` is architecture and 50% is state management. 
* link to separate MD file, so this doesn't disrupt flow

---

tag #3086
